### PR TITLE
fix training configs & trouble shooting

### PR DESCRIPTION
**Pull request by invictus717**
- Reviewed the running scripts and modified `sft.sh` for single-node execution, aligning its arguments with `sft_slurm.sh`.  

- Fixed issues encountered during reproduction, including dependencies, GPU memory usage, and runtime configurations. Relevant details are annotated in the script.  

- Added reference reproduction results to `sft.sh`.  

- Provided a quick-start guide for easier result reproduction.  

- Added a `.gitignore` file to exclude training artifacts.

### CODE REVIEW

File: train/sft.py [hunk 1]
```diff
@@ -16,7 +16,6 @@ class TrainingConfig:
     model_name: str = field(default="Qwen/Qwen2.5-32B-Instruct")
     block_size: int = field(default=32768)
     wandb_project: Optional[str] = field(default="s1")
-    wandb_entity: Optional[str] = field(default="hashimoto-group")
```
Comment (Muennighoff) [inline train/sft.py:19]: Won't this fail cuz it is called in __post_init__?

File: train/sft.sh [hunk 2]
```diff
@@ -1,37 +1,41 @@
...
+    --per_device_train_batch_size=1 \
+    --per_device_eval_batch_size=1 \
+    --gradient_accumulation_steps=1 \
+    --num_train_epochs=${epochs} \
+    --train_file_path="simplescaling/s1K_tokenized" \
+    --model_name="Qwen/Qwen2.5-32B-Instruct" \
+    --warmup_ratio=0.05 \
+    --report_to="none" \
+    --fsdp="full_shard auto_wrap" \
+    --fsdp_config="train/fsdp_config_qwen_cpu.json" \
```
Comment (Muennighoff) [inline train/sft.sh:26]: Maybe can leave this one at `train/fsdp_config_qwen.json`

File: train/sft.sh [hunk 3]
```diff
@@ -1,37 +1,41 @@
...
+    --eval_strategy="no" \
+    --logging_steps=1 \
+    --save_strategy="no" \
+    --lr_scheduler_type="cosine" \
+    --learning_rate=${lr} \
+    --weight_decay=${weight_decay} \
+    --adam_beta1=0.9 \
+    --adam_beta2=0.95 \
+    --output_dir="ckpts/${run_name}" \
+    --push_to_hub=false \
```
Comment (Muennighoff) [inline train/sft.sh:37]: Maybe can leave this one as being configurable

**Review by Muennighoff — COMMENTED** [review]
Nice!!

**Comment by invictus717** [comment]
Thank you for the review! It has now been updated.

File: train/sft.sh [hunk 4]
```diff
@@ -1,37 +1,41 @@
...
 uid="$(date +%Y%m%d_%H%M%S)"
 base_model="Qwen/Qwen2.5-32B-Instruct"
 lr=1e-5
 min_lr=0
 epochs=5
-micro_batch_size=2 # -> batch_size will be 16 if 8 gpus
-push_to_hub=true
-gradient_accumulation_steps=2
+weight_decay=1e-4 # -> the same training pipe as slurm_training
+micro_batch_size=1 # -> batch_size will be 8 if 8 gpus
```
Comment (Muennighoff) [inline train/sft.sh:9]:
```suggestion
micro_batch_size=1 # -> batch_size will be 16 if 16 gpus
```

(I thin we did 2 nodes; 16 gpus)

File: train/sft.sh (hunk 3 above)
Comment (Muennighoff) [reply train/sft.sh:37]: I think push_to_hub is still false? Maybe could be configurable

Comment (invictus717) [reply train/sft.sh:37]: Now it's been updated

File: train/sft.sh [hunk 5]
```diff
@@ -1,37 +1,42 @@
...
---output_dir="ckpts/s1_${uid}" \
---hub_model_id="simplescaling/s1-${uid}" \
---push_to_hub=${push_to_hub} \
---save_only_model=True
\ No newline at end of file
+    train/sft.py \
+    --block_size=32768 \
+    --per_device_train_batch_size=1 \
+    --per_device_eval_batch_size=1 \
+    --gradient_accumulation_steps=1 \
```
Comment (Muennighoff) [inline train/sft.sh:20]:
```suggestion
    --per_device_train_batch_size=${micro_batch_size} \
    --per_device_eval_batch_size=${micro_batch_size} \
    --gradient_accumulation_steps=${gradient_accumulation_steps} \
```

File: train/sft.sh [hunk 6]
```diff
@@ -1,37 +1,42 @@
...
+    --num_train_epochs=${epochs} \
+    --train_file_path="simplescaling/s1K_tokenized" \
+    --model_name="Qwen/Qwen2.5-32B-Instruct" \
```
Comment (Muennighoff) [inline train/sft.sh:23]:
```suggestion
    --model_name=${base_model} \
```

File: train/sft.sh [hunk 7]
```diff
@@ -1,37 +1,42 @@
...
+    --warmup_ratio=0.05 \
+    --report_to="none" \
```
Comment (Muennighoff) [inline train/sft.sh:25]:
```suggestion
```

Maybe removing this and instead adding a comment sth like
# To deactivate wandb add `--report_to="none" \`

File: train/sft.sh [hunk 8]
```diff
@@ -1,37 +1,42 @@
...
+    --fsdp="full_shard auto_wrap" \
+    --fsdp_config="train/fsdp_config_qwen.json" \
+    --bf16=True \
+    --eval_strategy="no" \
+    --logging_steps=1 \
+    --save_strategy="no" \
```
Comment (Muennighoff) [inline train/sft.sh:31]: Why no saving?

File: train/sft.sh [hunk 9]
```diff
@@ -1,37 +1,42 @@
...
+    --lr_scheduler_type="cosine" \
+    --learning_rate=${lr} \
+    --weight_decay=${weight_decay} \
+    --adam_beta1=0.9 \
+    --adam_beta2=0.95 \
+    --output_dir="ckpts/s1-${uid}" \
+    --push_to_hub=${push_to_hub} \
+    --save_only_model=True \
+    --gradient_checkpointing=True \
```
Comment (Muennighoff) [inline train/sft.sh:40]: Maybe we can also remove this one for now in favor of a comment saying that if you run OOM add this

File: train/sft.sh [hunk 10]
```diff
@@ -1,37 +1,42 @@
...
+    --accelerator_config='{"gradient_accumulation_kwargs": {"sync_each_batch": true}}'
```
Comment (Muennighoff) [inline train/sft.sh:41]: Interesting; did it fail without this?

File: train/sft.sh (hunk 8 above)
Comment (invictus717) [reply train/sft.sh:30]: saving the last checkpoint after training for quick reproduction

File: train/sft.sh (hunk 9 above)
Comment (invictus717) [reply train/sft.sh:39]: OK

Comment (invictus717) [reply train/sft.sh:39]: I will mention this in the readme.md

File: train/sft.sh (hunk 10 above)
Comment (invictus717) [reply train/sft.sh:40]: Enable gradient checkpointing for efficient memory usage with 8 H100 GPUs. Yes.

File: README.md [hunk 11]
```diff
@@ -191,6 +195,14 @@ response = tokenizer.batch_decode(generated_ids, skip_special_tokens=True)[0]
...
 To run training, you can find our script at `train/sft.py` which you can invoke via one of the `train/sft*sh` scripts which in turn you can launch via `train/launch.sh` if you are on a SLURM cluster (requires editing the file for your cluster setup).
 
+We provide a quick start to reproduce the results
+```
+git clone https://github.com/simplescaling/s1.git
+cd s1
+pip3 install -r requirements.txt
+bash train/sft.sh
+```
+*Note: If you encounter an out-of-memory (OOM) issue with 8 GPUs, consider enabling gradient checkpointing by adding the following line to your script: `--gradient_checkpointing=True `.*
```
Comment (Muennighoff) [inline README.md:205]:
```suggestion
*Note: If you encounter an out-of-memory (OOM) issue with 8 GPUs, consider enabling gradient checkpointing by adding the following line to your script: `--gradient_checkpointing=True`.*
```

File: train/sft.sh (hunk 10 above)
Comment (Muennighoff) [reply train/sft.sh:39]: Without the `--accelerator_config='{"gradient_accumulation_kwargs": {"sync_each_batch": true}}'` it failed?
